Add contrib importance pie chart

christinaexyou opened the pull request:
This PR is part 2 of addressing issue #114 by making the following changes:
1. Creates a pie chart depicting contributor importance 
2. Updates `queries/contributors_query.py` to query for contributor alias emails and returns the dataframe with an `email` column included
3. Updates chaoss/chaoss.py

cdolfi commented on pages/chaoss/visualizations/contrib_importance.py:
```diff
@@ -0,0 +1,251 @@
+
+    df = df[df["Action"].str.contains(action_type)]
+
+    df = df[~df['email'].isin(patterns)]
```
What about doing the bot filtering by the login? Not sure how many bots actually have emails linked to the account beyond a "githubnoreply" one. Also would be able to get it from the contributors actions query without any changes. I am also seeing if augur get bot tags (https://github.com/3scale/3scale-operator/pull/860) example of a bot with a bot tags, will luk. @JamesKunstle thoughts?

cdolfi commented on pages/chaoss/visualizations/contrib_importance.py:
```diff
@@ -0,0 +1,251 @@
+import random 
+
+PAGE = "chaoss"  
+VIZ_ID = "contrib-importance"  
```
change to "contrib-importance-pie" or something like it to make the naming descriptively distinct to the line graph (same for the file name)

cdolfi commented on pages/chaoss/visualizations/contrib_importance.py:
```diff
@@ -0,0 +1,251 @@
+        dbc.CardBody(
+            [
+                html.H3(
+                    "Contributor Importance by Action Type ",
```
A little more descriptive title maybe, @JamesKunstle thoughts?

cdolfi reviewed (changes requested):
A few requested changes. Going to look into the bot tagging!

cdolfi commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,248 @@
+                        dbc.PopoverBody("""
+                                        Visualizes the proportion of top k contributors and\n
+                                        remaining contributors according to a specific action\n
+                                        type. Option to filter out contributors with 'bot' in email.
```
change to be by log in not email

cdolfi replied:
Add details on the cntrb ids, (anonymized logins)

cdolfi commented on queries/contributors_query.py:
```diff
@@ -9,7 +9,6 @@
 
 QUERY_NAME = "CONTRIBUTOR"
 
-
```
Drop this file from being changed. Should be the same as before

cdolfi commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,248 @@
+    df = df[df["Action"].str.contains(action_type)]
+
+    # remove substring matches between login and patterns
+    df = df[~df['login'].isin(patterns)]
```
Should this be in the same format as line 200? I thought isin would require that df.login (in this example) string is in the patterns list, no the subset

christinaexyou replied:
Double checked and isin checks if df.login is an exact match and NOT substring matches to any element in the patterns list. Edited my comment  in line 208: `# remove rows where the login column value is in patterns - requires the login string to be in patterns`

cdolfi replied:
So this should be changed to .str.contains(patterns) as we are looking for substrings and that is what the search bar is prompting for

christinaexyou commented:
Changes made:

1. Addressed no data issue for PR Review and Issue Open 
2. Search for potential bots in login rather than email - this is mostly just a temp fix
3. Revert contributors_query.py to the most recent update in the dev branch

cdolfi commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,248 @@
+    # create plotly express pie chart
+    fig = px.pie(
+        df, 
+        names="cntrb_id", 
```
Add in comment that names could be changed to "login" to un- anonymize

christinaexyou replied:
Added the following comment in line 242: `# can be replaced with login to unanonymize`

cdolfi commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,248 @@
+        names="cntrb_id", 
+        values=action_type, 
+        color_discrete_sequence=color_seq, 
+        hole=.6)
```
Take out hole to be able to see splits better and match format of other pie charts

cdolfi commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,248 @@
+        dbc.CardBody(
+            [
+                html.H3(
+                    "Contributor Importance by Action Type ",
```
@JamesKunstle Curious to see if you have any ideas on naming of this graph, the line, and then the action types viz I made last week. Trying to think if there is any better naming to make them distinct or if its not really possible/ big deal

cdolfi commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,248 @@
+    fig.update_traces(
+        textinfo='percent+label',
+        textposition='inside',
+        hovertemplate='%{label} <br>Contributions: %{value}<br><extra></extra>')
```
Add "Contrib Id" before label or another identifier

cdolfi reviewed (changes requested):
Some small changes requested. Its getting close!

cdolfi commented:
@christinaexyou is this meant to be closed?

cdolfi commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,256 @@
+                                        type. Contributors who are in the top k of contributions\n 
+                                        are anonynimzed by their unique alpha-numeric Contributor ID.\n
+                                        The remaining contributors are labeled as 'other'. Option to\n
+                                        filter out contributors that have 'bot' as their login username.
```
Add "Note: Some commits have a contributor ID as none if there is no github account associated with the email used to commit"

christinaexyou replied:
Added note - replaced "commit" to "contributions" to generalize

cdolfi reviewed (changes requested):
One small additional review comment. I resolved all of the prior comments that have been addressed. Once the rest are handled this will be good to go!

christinaexyou commented:
The following changes were made:
1. Corrected original comment about `isin` for removing bots - it looks for exact matches between the login and the elements in patterns NOT substring matches
2. Updated graph info with a note about contributors that are ID'ed as None

cdolfi reviewed (changes requested):
Updated review comments

cdolfi commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,262 @@
+    # option to filter out potential bots
+    if patterns:
+        # remove rows where Login column value contains the substring 'bot'
+        df = df[df['login'].str.contains(str(patterns)[1:-1])==False]
```
Use ~ not ==False and use the following logic: https://www.skytowner.com/explore/getting_rows_where_column_value_contains_any_substring_in_a_list_in_pandas_dataframe

christinaexyou replied:
Included `na=False` clause in mask variable since there are some nan/empty login column values.

cdolfi replied:
Have you specifically check if using ~ works for nan and empty values? Every check ive done shows that it would

christinaexyou replied:
When I just do ~, it returns `TypeError: bad operand type for unary -: 'NoneType' ` so I referenced this post: https://stackoverflow.com/questions/52297740/typeerror-bad-operand-type-for-unary-float

cdolfi reviewed (changes requested):
One last thing and then just waiting on @JamesKunstle 's review

cdolfi reviewed (commented):
lgtm! Ready to merge once @JamesKunstle has a chance to take a look at the few comments left

JamesKunstle commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,263 @@
+                                        login username. Note: some commits have a Contributor ID\n
+                                        as none if there is no Github account associated with the\n
+                                        email used to commit.
+                                        """),
```
Here's a slight reframe suggestion:

"For a given kind of contribution (e.g. Commit, PR, Issue), visualizes the proportional share of the top 'k' anonymous contributors, aggregating the remaining contributors as "other".

e.g. Contributor 'A' opens the most PRs of all contributors, accounting for 1/5 of all PRs. If 'k' = 1, then the chart will have one slice for Contributor 'A' accounting for 1/5 of the area, with the remaining 4/5 representing all other contributors.

Optionally, filter out contributors who have 'bot' in their login username.

Note: Some commits may have a Contributors ID as 'none' if no Github account is associated with the email that the contributor committed as."

christinaexyou replied:
What about "Top {k} Contributors by {action}" where k and action are dynamically changing based on the user input ? For the pie chart, I'm not explicitly calculating bus factor - just visualizing the data as is. I like the reframe suggestion - will update `Graph Info`.

cdolfi replied:
@JamesKunstle

JamesKunstle replied:
@christinaexyou @cdolfi Yeah "Top {k} Contributors by Number of {action}" might be slightly clearer but we're within a razors margin of a perfect title, to me, either way, so I'm down for your preference.

cdolfi replied:
lgtm on info and graph title, will leave for @JamesKunstle

JamesKunstle commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,263 @@
+    df = df.rename(columns={'Action': action_type})
+
+    # get the number of total contributions 
+    t_sum = df.iloc[:,1].sum()
```
in Pandas, it's preferred to use Series methods rather than DataFrame indexing, so I believe this line would be replaced with:

`t_sum = df[action_type].sum()`

[I think.](https://www.geeksforgeeks.org/python-pandas-series-sum/#)

JamesKunstle commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,263 @@
+    t_sum = df.iloc[:,1].sum()
+
+    # index df to get first k rows and remove 'cntrbs' column
+    df = df.iloc[:top_k, :2]
```
Same kind of semantic Pandas thing here:

`df = df.head(top_k).drop('cntrbs')`

https://sparkbyexamples.com/pandas/get-first-n-rows-of-pandas/#:~:text=DataFrame.,rows%20from%20the%20given%20DataFrame.

JamesKunstle commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,263 @@
+    df['cntrb_id'] = df['cntrb_id'].apply(lambda x: str(x).split('-')[0])
+    
+    # sort by alpha-numeric order to make it easier to map colors to contibutors
+    df_sum = df.iloc[:,1].sum() # get the number of total top k contributions
```
Same kind of semantic Pandas thing, again, because I'm picky:

`df_sum = df[action_type].sum()`

I assume this is for the action column.

JamesKunstle commented on pages/chaoss/visualizations/contrib_importance_pie.py:
```diff
@@ -0,0 +1,263 @@
+    df_sum = df.iloc[:,1].sum() # get the number of total top k contributions
+    
+    # calculate the remaining contributions by taking the the difference of t_sum and df_sum
+    df.loc[len(df.index)] = ['other', t_sum - df_sum ]
```
so me, adding a row should be refactored to:

`df = df.append({"cntrb_id": "Other", action_type: str(t_sum - df_sum)})`

because the `df.loc[len(df.index)]` hack is difficult for a novice to read.

JamesKunstle reviewed (changes requested):
Minor semantic Pandas updates requested but nothing technical- this seems very good.

I added a recommendation for the 'about' section and the title- please let me know what you think.

christinaexyou commented:
@cdolfi @JamesKunstle 
Changes made:

1. Added functionality user to create new keywords to filter out contributors by 
2. Create dynamically changing title
3. Updated `Graph Info` description
4. Edited Pandas logic

cdolfi reviewed (approved):
@JamesKunstle to do final review but lgtm

JamesKunstle reviewed (approved):
I think this is done. Good work!
